backends/ebpf/ebpfenv.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+# Copyright 2018 VMware, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+""" Virtual environment which models a simple bridge with n attached
+    interfaces. The bridge runs in a completely isolated namespace.
+    Allows the loading and testing of eBPF programs by initializing every
+    interface with a qdisc. """
+
+import os
+import sys
+from subprocess import Popen, PIPE
+sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)) + '/../../tools')
+from testutils import *
+
+
+class Bridge(object):
+
+    def __init__(self, namespace, outputs, verbose):
+        self.ns_name = namespace     # identifier of the namespace
+        self.br_name = "core"        # name of the central bridge
+        self.br_ports = []           # list of bridge ports
+        self.outputs = outputs       # contains standard and error output
+        self.verbose = verbose       # do we want to be chatty?
+
+    def ns_init(self):
+        """ Initialize the namespace. """
+        cmd = "ip netns add %s" % self.ns_name
+        errmsg = "Failed to create namespace %s :" % self.ns_name
+        result = run_timeout(True, cmd, TIMEOUT,
+                             self.outputs, errmsg)
+        self.ns_exec("ip link set dev lo up")
```

Comment:
is there a problem if the previous command failed and you try to run this one?

Reply:
Not realy, the ns_exec cmd will return an error, but I technically do not care about the output and it will just return the error of the previous command.